src/Samples/Sarif.WorkItems.Sample/Program.cs

```diff
@@ -5,112 +5,83 @@
-            return 0;
-        }
-
-        private static void NonFunctioningEndToEndApiUsageExample()
-        {
+            // Mock up of end-to-end API use.
+            
             // 1. Retrieve the current SARIF and its baseline equivalent
-            SarifLog baselineSarif = RetrieveBaselineSarif();
-            SarifLog currentSarif = RetrieveCurrentSarif();
+            SarifLog baselineSarif = RetrieveBaselineSarif(args[0]);
+            SarifLog currentSarif = RetrieveCurrentSarif(args[1]);
+            string outputFilePath = args[2];
 
-            // 2. Match the SARIF files
+            // 2. Given a baseline and current sarif, produce a 'matched' log
             ISarifLogBaseliner resultMatcher = SarifLogBaselinerFactory.CreateSarifLogBaseliner(SarifBaselineType.Standard);
 
-            SarifLog matchedLog = new SarifLog
-            {
-                Runs = new[]
-                {
-                    resultMatcher.CreateBaselinedRun(baselineSarif.Runs[0], currentSarif.Runs[0])
-                }
-            };
+            SarifLog matchedSarif = resultMatcher.CreateBaselinedSarifLog(
+                baselineSarif, 
+                currentSarif);
 
-            // 3. Make sure everything has a GUID and that we have a map back to each result from it
-            var guidToMatchedResultMap = new Dictionary<string, Result>();
-            foreach (Result result in matchedLog.Runs[0].Results)
-            {
-                if (result.Guid == null)
-                {
-                    result.Guid = Guid.NewGuid().ToString();
-                }
-                guidToMatchedResultMap[result.Guid] = result;
-            }
+            // 3. We need a map from each result guid to the matched SARIF result.
+            //    As bugs are filed against refactored files, we use the guid to 
+            //    link to the original file in order to populate the WIT URI.
 
-            var adoClient = new AzureDevOpsClient();
-            var filingTarget = new AzureDevOpsFilingTarget(adoClient);
+            var guidMappingVisitor = new GuidMappingVisitor();
+            guidMappingVisitor.Visit(matchedSarif);
 
-            var workItemFiler = new WorkItemFiler(
-                filingTarget,
-                filteringStrategy: new NewResultsFilteringStrategy(),
-                groupingStrategy: new OneResultPerWorkItemGroupingStrategy());
+            // 4. Now we need to refactor the log file into one or more log files, 
+            // each of which will drive the creation of a single bug.
 
-            string logFilePath = Path.GetTempFileName();
-            File.WriteAllText(logFilePath, JsonConvert.SerializeObject(matchedLog));
+            IEnumerable<SarifLog> sarifLogsToFile = SarifSplitter.Split(
+                sarifLog: matchedSarif,
+                filterPredicate: (result) => { return result.BaselineState == BaselineState.New; },
+                splittingStrategy: SplittingStrategy.OneFilePerRunPerRule);
```

Comment:
>splittingStrategy [](start = 16, length = 17)

What was your reasoning in using a bare function for `filterPredicate`, but a strategy class (with a single method) for `splittingStrategy`?

I like the strategy class approach because you can add more methods or properties later on without breaking existing callers. (At least, you can do that if all your strategy classes derive from an abstract base class, as the ones in my implementation do.)

Reply:
Bottom line, I like:
```
filterStrategy: FilteringStrategies.NewResults,
splittingStrategy: SplittingStrategies.OneFilePerRunPerRule
```
And I like `FilteringStrategies`/`SpilttingStrategies` (plural) so I can reserve `FilteringStrategy`/`SplittingStrategy` as the name of the abstract base class.

OTOH, I suppose you could put those prebuilt objects _in_ the abstract base class, but I don't like the idea of making the ABC aware of all its descendants. Let someone else have that knowledge.

---
In reply to: [307084776](https://github.com/microsoft/sarif-sdk/pull/1589#discussion_r307084776) [](ancestors = 307084776)

Reply:
Yes, it has to be that way. If someone writes their own strategy and derives from the ABC, you don't want them to have to carry along the dependencies (if any) of all the out-of-the-box concrete classes. Ok, fine, they are probably dragging those dependencies along anyway because the same assembly that contains the ABC (e.g. `FilteringStrategy` also contains the static holder `FilteringStrategies`. So I guess that's not a strong argument. I just don't like the idea that a class knows its descendants.

---
In reply to: [307085292](https://github.com/microsoft/sarif-sdk/pull/1589#discussion_r307085292) [](ancestors = 307085292,307084776)

Reply:
Filtering is a predicate because filtering is naturally a true/false operation (is this thing in or out)? Splitting is entirely different because you are reorganizing multiple sets into arbitrary other sets. I don't see filtering as benefiting from object inheritance. Someone could create a static predicate instance that bolts together other static predicate instances though. I need to think more about strategy injection into this splitter class. The current design simply provides an enum to drive the specific operation you want. Again, the reorganization that's implied here is so arbitrary that I can't see the need for layering. i.e., you basically want an object with a Split() method that accepts one log file and produces many. From this angle, the API should actually instantiate and invoke one of these concrete types, rather than parameterizing. Let me think a bit more.

---
In reply to: [307089548](https://github.com/microsoft/sarif-sdk/pull/1589#discussion_r307089548) [](ancestors = 307089548,307085292,307084776)